### PR TITLE
serde: implement `serialize_pointers`

### PR DESCRIPTION
**Pull request by tcoratger**

### CODE REVIEW

File: crates/exex/src/serde.rs [hunk 1]
```diff
@@ -105,7 +146,7 @@ mod tests {
 
     fn setup_kakarot_serde() -> KakarotSerde {
         // Load the valid program content from a JSON file
-        let program_content = include_bytes!("../../../cairo/programs/os.json");
+        let program_content = include_bytes!("../testdata/os.json");
```
Comment (tcoratger) [inline crates/exex/src/serde.rs:149]: @ClementWalter I know that here you preferred that I directly use the OS program located in the Cairo part. But I isolated it in the Rust part for the following reasons:
- I am currently testing the serialization functions in a unitary way (without complete serialization of a realistic program).
- So I mock almost everything and I base myself only on the JSON (not on a complete Cairo logic, I will make at the end a complete test which will be "Cairo readable" to deserialize).
- As I base myself on these things so low level, as soon as we modify a single line of the Cairo OS and therefore its JSON, all our Rust unit tests will explode and it will take a crazy amount of time to fix them. 
- So basically this JSON file = we must never touch it

File: crates/exex/src/serde.rs [hunk 2]
```diff
@@ -223,4 +264,108 @@ mod tests {
...
+
+        // Try to serialize pointer with main which is not a struct.
+        if let Err(KakarotSerdeError::IdentifierNotFound { struct_name, expected_type }) =
+            kakarot_serde.serialize_pointers("main", Relocatable { segment_index: 0, offset: 0 })
+        {
+            assert_eq!(struct_name, "main".to_string());
+            assert_eq!(expected_type, Some("struct".to_string()));
+        } else {
+            panic!("Expected KakarotSerdeError::IdentifierNotFound");
+        }
```
Comment (ClementWalter) [inline crates/exex/src/serde.rs:284]:
```suggestion
        panic!("Expected KakarotSerdeError::IdentifierNotFound");
```

why do we need a if/else?

File: crates/exex/src/serde.rs [hunk 3]
```diff
@@ -223,4 +264,108 @@ mod tests {
...
+            .vm
+            .insert_value(new_segment, Relocatable { segment_index: 0, offset: 0 });
+
+        // Loop over a range of values to insert relocatable values into memory.
+        for i in 1..7 {
+            let _ = kakarot_serde.runner.vm.insert_value(
+                (new_segment + i as usize).unwrap(),
+                Relocatable { segment_index: 10, offset: 11 },
+            );
+        }
```
Comment (ClementWalter) [inline crates/exex/src/serde.rs:309]: use `load_data` or `gen_arg` https://github.com/lambdaclass/cairo-vm/blob/main/vm/src/vm/vm_memory/memory_segments.rs#L132

File: crates/exex/src/serde.rs (hunk 2 above)
Comment (tcoratger) [reply crates/exex/src/serde.rs]: Fixed, just used a match to be able to check the validity of the error message

**Review by ClementWalter — COMMENTED** [review]
if you want to start from a fixed small cairo or json, you can create a minimal cairo file and compile it like

```
uv run compile cairo/tests/utils/serde.cairo
```

File: crates/exex/src/serde.rs [hunk 4]
```diff
@@ -223,4 +264,108 @@ mod tests {
...
+        // Add a new memory segment to the virtual machine (VM).
+        let new_segment = kakarot_serde.runner.vm.add_memory_segment();
+
+        // Adding new zero values to check the effect of pointers vs non pointers
+        for i in 0..2 {
+            let _ = kakarot_serde.runner.vm.insert_value(
+                (new_segment + i as usize).unwrap(),
+                Relocatable { segment_index: 0, offset: 0 },
+            );
+        }
```
[the remark this answers is not in the document]
Comment (tcoratger) [reply crates/exex/src/serde.rs]: fixed

File: crates/exex/src/serde.rs [hunk 5]
```diff
@@ -223,4 +264,98 @@ mod tests {
...
+            Relocatable { segment_index: 10, offset: 11 },
+        ]);
+
+        // Serialize the pointers of the "ImplicitArgs" struct using the new memory segment.
+        let result = kakarot_serde
+            .serialize_pointers(
+                "apply_transactions.ImplicitArgs",
+                Relocatable { segment_index: 0, offset: 0 },
+            )
+            .expect("failed to serialize pointers");
```
Comment (ClementWalter) [inline crates/exex/src/serde.rs:312]:
```suggestion
        let base = kakarot_serde.runner.vm.gen_arg(&vec![
            Relocatable { segment_index: 10, offset: 11 },
            Relocatable { segment_index: 10, offset: 11 },
            Relocatable { segment_index: 10, offset: 11 },
            Relocatable { segment_index: 10, offset: 11 },
            Relocatable { segment_index: 10, offset: 11 },
            Relocatable { segment_index: 10, offset: 11 },
        ]);

        // Serialize the pointers of the "ImplicitArgs" struct using the new memory segment.
        let result = kakarot_serde
            .serialize_pointers(
                "apply_transactions.ImplicitArgs",
                base,
            )
            .expect("failed to serialize pointers");
```

File: crates/exex/src/serde.rs [hunk 6]
```diff
@@ -223,4 +264,98 @@ mod tests {
...
+            Relocatable { segment_index: 0, offset: 0 },
+        ]);
+
+        // Try to serialize
+        let result = kakarot_serde
+            .serialize_pointers(
+                "apply_transactions.Args",
+                Relocatable { segment_index: 0, offset: 0 },
+            )
+            .expect("failed to serialize pointers");
```
Comment (ClementWalter) [inline crates/exex/src/serde.rs:349]:
```suggestion
        let base = kakarot_serde.runner.vm.gen_arg(&vec![
            Relocatable { segment_index: 0, offset: 0 },
            Relocatable { segment_index: 0, offset: 0 },
        ]);

        // Try to serialize
        let result = kakarot_serde
            .serialize_pointers(
                "apply_transactions.Args",
                base,
            )
            .expect("failed to serialize pointers");
```

Comment (tcoratger) [reply crates/exex/src/serde.rs]: fixed

File: crates/exex/src/serde.rs (hunk 5 above)
Comment (tcoratger) [reply crates/exex/src/serde.rs]: fixed

**Comment by tcoratger** [comment]
> if you want to start from a fixed small cairo or json, you can create a minimal cairo file and compile it like
> 
> ```
> uv run compile cairo/tests/utils/serde.cairo
> ```

fixed

File: crates/exex/src/serde.rs [hunk 7]
```diff
@@ -218,9 +261,102 @@ mod tests {
...
+        // Assert that the result matches the expected serialized struct members.
+        assert_eq!(
+            result,
+            HashMap::from_iter([
+                ("bitwise_ptr".to_string(), Some(Relocatable { segment_index: 10, offset: 11 })),
+                ("output_ptr".to_string(), None),
+                (
+                    "range_check_ptr".to_string(),
+                    Some(Relocatable { segment_index: 10, offset: 11 })
+                ),
```
Comment (ClementWalter) [inline crates/exex/src/serde.rs:323]:
```suggestion
                ("output_ptr".to_string(), None),
                (
                    "range_check_ptr".to_string(),
                    Some(Relocatable { segment_index: 10, offset: 11 })
                ),
                ("bitwise_ptr".to_string(), Some(Relocatable { segment_index: 10, offset: 11 })),                
```

Comment (ClementWalter) [reply crates/exex/src/serde.rs:323]: to preserve order

File: crates/exex/src/serde.rs [hunk 8]
```diff
@@ -218,9 +261,102 @@ mod tests {
...
+            .expect("failed to serialize pointers");
+
+        // Assert that the result matches the expected serialized struct members.
+        assert_eq!(
+            result,
+            HashMap::from_iter([
+                ("bitwise_ptr".to_string(), Some(Relocatable { segment_index: 10, offset: 11 })),
+                // Not a pointer so that we shouldn't have a `None`
+                ("range_check_ptr".to_string(), Some(Relocatable { segment_index: 0, offset: 0 })),
+                ("output_ptr".to_string(), Some(Relocatable { segment_index: 10, offset: 11 })),
```
Comment (ClementWalter) [inline crates/exex/src/serde.rs:358]:
```suggestion
                ("output_ptr".to_string(), Some(Relocatable { segment_index: 10, offset: 11 })),
                ("range_check_ptr".to_string(), Some(Relocatable { segment_index: 0, offset: 0 })),                
                ("bitwise_ptr".to_string(), Some(Relocatable { segment_index: 10, offset: 11 })),
```

Comment (tcoratger) [reply crates/exex/src/serde.rs]: fixed

**Review by ClementWalter — APPROVED** [review]
